shared_model/backend/protobuf/transaction.hpp

```diff
@@ -0,0 +1,125 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef IROHA_SHARED_MODEL_PROTO_TRANSACTION_HPP
+#define IROHA_SHARED_MODEL_PROTO_TRANSACTION_HPP
+
+#include "interfaces/transaction.hpp"
+
+#include <boost/range/numeric.hpp>
+
+#include "backend/protobuf/commands/proto_command.hpp"
+#include "backend/protobuf/common_objects/signature.hpp"
+#include "backend/transport_getter.hpp"
+#include "block.pb.h"
+#include "utils/lazy_initializer.hpp"
+#include "utils/reference_holder.hpp"
+
+namespace shared_model {
+  namespace proto {
+    class Transaction final
+        : public interface::Transaction,
+          public TransportGetter<iroha::protocol::Transaction> {
+     public:
+      template <typename TransactionType>
+      explicit Transaction(TransactionType &&transaction)
+          : transaction_(std::forward<TransactionType>(transaction)),
+            payload_([this]() -> decltype(auto) {
+              return (transaction_->payload());
+            }),
+            commands_([this] {
+              return boost::accumulate(payload_->commands(),
+                                       CommandsType{},
+                                       [](auto &&acc, const auto &cmd) {
+                                         acc.emplace_back(new Command(cmd));
+                                         return acc;
+                                       });
+            }),
+            blob_([this] { return BlobType(payload_->SerializeAsString()); }),
+            signatures_([this] {
+              return boost::accumulate(transaction_->signature(),
+                                       SignatureSetType{},
+                                       [](auto &&acc, const auto &sig) {
+                                         acc.emplace(new Signature(sig));
+                                         return acc;
+                                       });
+            }) {}
+
+      Transaction(const Transaction &o) : Transaction(*o.transaction_) {}
+
+      Transaction(Transaction &&o) noexcept
+          : Transaction(std::move(o.transaction_.variant())) {}
+
+      const interface::types::AccountIdType &creatorAccountId() const override {
+        return payload_->creator_account_id();
+      }
+
+      TxCounterType transactionCounter() const override {
+        return payload_->tx_counter();
+      }
+
+      const CommandsType &commands() const override { return *commands_; }
+
+      ModelType *copy() const override {
+        return new Transaction(iroha::protocol::Transaction(*transaction_));
+      }
+
+      const BlobType &blob() const override { return *blob_; }
+
+      const SignatureSetType &signatures() const override {
+        return *signatures_;
+      }
+
+      bool addSignature(const SignatureType &signature) override {
+        if (signatures_->count(signature) > 0) {
+          return false;
+        }
+        auto sig = transaction_->add_signature();
```

Comment:
auto [discards](http://en.cppreference.com/w/cpp/language/template_argument_deduction) reference

Reply:
That is fine, since `add_signature` returns pointer.

Reply:
Ah, you're right, missed it